core/src/Kokkos_Swap.hpp

```diff
@@ -0,0 +1,67 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOS_SWAP_HPP
+#define KOKKOS_SWAP_HPP
+
+#include <Kokkos_Macros.hpp>
+
+#include <cstddef>
+#include <type_traits>
+#include <utility>
+
+namespace Kokkos {
+
+template <class T>
+KOKKOS_FUNCTION constexpr std::enable_if_t<std::is_move_constructible_v<T> &&
+                                           std::is_move_assignable_v<T>>
+swap(T& a, T& b) noexcept(std::is_nothrow_move_constructible_v<T>&&
+                              std::is_nothrow_move_assignable_v<T>) {
+  T t(std::move(a));
+  a = std::move(b);
+  b = std::move(t);
+}
+
+namespace Impl {
+
+template <class T>
+struct is_swappable {
+  template <class U>
+  static decltype(swap(std::declval<T&>(), std::declval<T&>())) test_swap(int);
+  struct Nope;
+  template <class U>
+  static Nope test_swap(long);
+  static constexpr bool value =
+      !std::is_same_v<decltype(test_swap<T>(0)), Nope>;
+};
```

Comment:
is this copied from the std more or less?

Reply:
The standard implements `is_swappable` but we can't really use it because it would not add `Kokkos::swap` as one of the possible overloads (unless the arguments belong to namespace Kokkos::).
But yes more or less equivalent to what you'd do in the standard implementation.